Core/Src/main.c

```diff
@@ -220,41 +258,42 @@ void task_main(void *argument)
       if (can_send(hGS_CAN.channels[frame_object.frame.channel], &frame_object.frame)) {
         /* Echo sent frame back to host */
         frame_object.frame.reserved = 0x0;
-        xQueueSendToBack(queue_to_hostHandle, &frame_object.frame, 0);
+        xQueueSendToBack(hGS_CAN.queue_to_hostHandle, &frame_object.frame, 0);
         can_on_tx_cb(frame_object.frame.channel, &frame_object.frame);
       }
       else {
         /* throw the message back onto the queue */
-        xQueueSendToFront(queue_from_hostHandle, &frame_object.frame, 0);
+        xQueueSendToFront(hGS_CAN.queue_from_hostHandle, &frame_object.frame, 0);
+        vTaskDelay(pdMS_TO_TICKS(0));
       }
     }
+  }
+}
 
+/**
+  * @brief  Function implementing thread for handling messages TO host.
+  * @param  argument: Not used
+  * @retval None
+  */
+void task_queue_to_host(void *argument)
+{
+  UNUSED(argument);
+  struct gs_host_frame_object frame_object;
+    
+    /* Infinite loop */
+  for(;;)
+  { 
     /* Check the queue to see if we have data TO the host to handle */
-    if (xQueueReceive(queue_to_hostHandle, &frame_object.frame, 0) == pdPASS) {
+    if (xQueueReceive(hGS_CAN.queue_to_hostHandle, &frame_object.frame, portMAX_DELAY) == pdPASS) {
       if (USBD_GS_CAN_SendFrame(&hUSB, &frame_object.frame) == USBD_OK) {
         can_on_rx_cb(frame_object.frame.channel, &frame_object.frame);
       }
       else {
         /* throw the message back onto the queue */
-        xQueueSendToFront(queue_to_hostHandle, &frame_object.frame, 0);
+        xQueueSendToFront(hGS_CAN.queue_to_hostHandle, &frame_object.frame, 0);
+        vTaskDelay(pdMS_TO_TICKS(0));
```

Comment:
How long does it actually wait here?

Reply:
I actually changed this to 1 in a later commit.  With 0 it simply goes back to the idle task and checks for lower priority tasks.  The issue with 0 is that since there are lower priority tasks (the main task) it would immediately retrigger the "to host" task as there was still data in the queue.  So it would be blocking until the queue was clear.  Changing it to 1 allows lower priority tasks to run and allow the USB xmit buffer to flush.

Reply:
Haven't wrapped my head around it, but if you rely on a sleep, the priorities are not setup correctly. The RX path from CAN to Host is the most critical path, as we don't have any flow control on RX.

Reply:
It's a holdoff if the SendFrame returns as busy when the TxQueue flag = 1 (the DataIn callback hasn't fired yet after the last TX).  We can always raise the priority of this task where an entry into the device->host queue will immediately trigger the task with preemption.  The issue I've seen is that the bottleneck is the SendFrame returns "busy" a LOT.